File: algorithms/solutions/week29/toanha/count-and-say.go
package main

import "fmt"

func countAndSay(n int) string {
	if n == 1 {
		return "1"
	}

	s := countAndSay(n - 1)
	result := ""
	count := 0

	for i := 0; i < len(s); i++ {
		count++
		if i == len(s)-1 || s[i] != s[i+1] {
			result += fmt.Sprintf("%d", count)
			result += fmt.Sprintf("%d", s[i]-'0')
			count = 0
		}
	}

	return result
}

func main() {
	fmt.Println(string(countAndSay(4)))
}

File: algorithms/solutions/week29/unrealhoang/count_and_say.rb
def count(s)
  last = nil
  last_count = 0
  result = []
  s.chars.each do |i|
    if last != i
      if last_count > 0
        result << last_count
        result << last
      end
      last = i
      last_count = 1
    else
      last_count += 1
    end
  end
  if last_count > 0
    result << last_count
    result << last
  end
  result.join('')
end

# @param {Integer} n
# @return {String}
def count_and_say(n)
  t = "1"
  (n-1).times do |i|
    t = count(t)
  end
  t
end
